solidity/test/Coordinator_test.js

```diff
@@ -37,133 +45,117 @@ contract('Coordinator', () => {
-        /// / so we have some indirect test coverage.
-        /// /
-        /// / https://github.com/ethereum/web3.js/issues/1241
-        /// / assert.equal(
-        /// /   sa[2],
-        /// /   ['0x70AEc4B9CFFA7b55C0711b82DD719049d615E21d', '0xd26114cd6EE289AccF82350c8d8487fedB8A0C07']
-        /// / )
+        initiateServiceAgreement(coordinator, defaultArgs)
+        checkServiceAgreementPresent(coordinator, serviceAgreementID, defaultArgs)
       })
     })
 
     context("with an invalid oracle signatures", () => {
-      beforeEach(() => {
-        oracleSignature = personalSign(newAddress(stranger), serviceAgreementID)
-        const requestDigestAddr = recoverPersonalSignature(serviceAgreementID, oracleSignature)
-        assert.notEqual(toHex(oracle), toHex(requestDigestAddr))
+      const badOracleSignature =
+            personalSign(newAddress(stranger), serviceAgreementID)
+      const badRequestDigestAddr =
+            recoverPersonalSignature(serviceAgreementID, badOracleSignature)
+      assert.notEqual(toHex(oracle), toHex(badRequestDigestAddr))
+      it('saves no service agreement struct, if signatures invalid', async () => {
+        assertActionThrows(
+          async () =>
+            initiateServiceAgreement(
+              coordinator,
+              Object.assign(defaultArgs, { oracleSignature: badOracleSignature})))
+        checkServiceAgreementAbsent(coordinator, serviceAgreementID)
       })
-
-      it('saves a service agreement struct from the parameters', async () => {
-        assertActionThrows(async () => {
-          await coordinator.initiateServiceAgreement(
-            toHex(payment),
-            toHex(expiration),
-            oracles.map(toHex),
-            [oracleSignature.v],
-            [oracleSignature.r].map(toHex),
-            [oracleSignature.s].map(toHex),
-            toHex(requestDigest)
-          )
-        })
-
-        const sa = await coordinator.serviceAgreements.call(toHex(serviceAgreementID))
-        assertBigNum(sa[0], bigNum(0))
-        assertBigNum(sa[1], bigNum(0))
-        assert.equal(sa[2], '0x0000000000000000000000000000000000000000000000000000000000000000')
+    })
+    context("Validation of service agreement deadlines", () => {
+      it('Rejects a service agreement with an endAt date in the past', async () => {
+        assertActionThrows(
+          async () => initiateServiceAgreement(
+            coordinator,
+            Object.assign(defaultArgs, { endAt: newHash('1000')})))
+        checkServiceAgreementAbsent(coordinator, serviceAgreementID)
       })
     })
   })
 
   describe('#executeServiceAgreement', () => {
+    let tx, log
     const fHash = functionSelector('requestedBytes32(bytes32,bytes32)')
     const to = '0x80e29acb842498fe6591f020bd82766dce619d43'
     const payment = 1000000000000000000
-    let sAID, tx, log
+    const requestDigest = newHash('0x9ebed6ae16d275059bf4de0e01482b0eca7ffc0ffcc1918db61e17ac0f7dedc8')
```

Comment:
On the topic of readability, maybe we should just make this some string which is clearly a fixture value instead of a possibly meaningful magic number.

Reply:
For some reason, github is not presenting me with a response textbox for your [comment about the ABI encoding library](https://github.com/smartcontractkit/chainlink/pull/672#discussion_r228161583), so I'll respond here. Thanks, that looks like a better solution than what I've done here. I'll fix that up.

This idea with the `requestDigest` values is good, too. I've changed them to `0xdeadbeef...`.